### PR TITLE
[CUDA] Initial work for boosting and evaluation with CUDA

### PR DESCRIPTION
**Pull request by shiyu1994**
This PR initiates adding boosting objectives and evaluation metrics for `cuda_exp` (see #5163). This PR only creates interfaces for CUDA objectives and metrics, and default back to boosting and evaluation on CPU.

### CODE REVIEW

**Comment by shiyu1994** [comment]
@guolinke @jameslamb @StrikerRUS Could you kindly help to review this PR? After merging this one, we can start adding more objectives for `cuda_exp`. That will make the single-GPU version of `cuda_exp` complete. Thanks a lot~

File: src/boosting/cuda/cuda_score_updater.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+#include "cuda_score_updater.hpp"
+
+#ifdef USE_CUDA_EXP
+
+namespace LightGBM {
+
+CUDAScoreUpdater::CUDAScoreUpdater(const Dataset* data, int num_tree_per_iteration, const bool boosting_on_cuda):
+  ScoreUpdater(data, num_tree_per_iteration), num_threads_per_block_(1024), boosting_on_cuda_(boosting_on_cuda) {
```
Comment (guolinke) [inline src/boosting/cuda/cuda_score_updater.cpp:13]: is `boosting_on_cuda` always true? should we keep this variable?

File: src/treelearner/cuda/cuda_single_gpu_tree_learner.hpp [hunk 2]
```diff
@@ -118,6 +118,15 @@ class CUDASingleGPUTreeLearner: public SerialTreeLearner {
...
   score_t* cuda_hessians_;
+  /*! \brief whether boosting is done on CUDA */
+  const bool boosting_on_cuda_;
+
+  #ifdef DEBUG
+  /*! \brief gradients on CPU */
+  std::vector<score_t> host_gradients_;
+  /*! \brief hessians on CPU */
+  std::vector<score_t> host_hessians_;
+  #endif  // DEBUG
```
Comment (guolinke) [inline src/treelearner/cuda/cuda_single_gpu_tree_learner.hpp:129]: are these codes used?

File: src/treelearner/cuda/cuda_data_partition.cu [hunk 3]
```diff
@@ -1045,11 +1045,11 @@ __global__ void AddPredictionToScoreKernel(
       const data_size_t global_data_index = data_indices_in_leaf[local_data_index];
       const int leaf_index = cuda_data_index_to_leaf_index[global_data_index];
       const double leaf_prediction_value = leaf_value[leaf_index];
-      cuda_scores[local_data_index] = leaf_prediction_value;
+      cuda_scores[global_data_index] += leaf_prediction_value;
```
Comment (guolinke) [inline src/treelearner/cuda/cuda_data_partition.cu:1048]: this is bug ?

File: src/metric/metric.cpp [hunk 4]
```diff
@@ -14,53 +14,130 @@
...
+    } else if (type == std::string("gamma")) {
+      return new GammaMetric(config);
+    } else if (type == std::string("gamma_deviance")) {
+      return new GammaDevianceMetric(config);
+    } else if (type == std::string("tweedie")) {
+      return new TweedieMetric(config);
+    }
+  #ifdef USE_CUDA_EXP
   }
+  #endif  // USE_CUDA_EXP
```
Comment (guolinke) [inline src/metric/metric.cpp:140]: can we reduce these duplicated codes? like use predefined macros.

**Comment by StrikerRUS** [comment]
> That will make the single-GPU version of `cuda_exp` complete. 

I think you forgot about 
- #5086 
- #4781 
- #5062 
- #4785

**Comment by jameslamb** [comment]
> please kindly help to review this PR

Before we put any time into this PR, can you please help with https://github.com/microsoft/LightGBM/pull/5287#issuecomment-1159597587?

File: src/boosting/cuda/cuda_score_updater.cpp (hunk 1 above)
Comment (shiyu1994) [reply src/boosting/cuda/cuda_score_updater.cpp:13]: No. This is a work around when there are some objective functions not implemented by the CUDA version yet. See line 98 of gbdt.cpp in this PR.

File: src/treelearner/cuda/cuda_single_gpu_tree_learner.hpp (hunk 2 above)
Comment (shiyu1994) [reply src/treelearner/cuda/cuda_single_gpu_tree_learner.hpp:129]: Yes. These are used in DEBUG mode to check that the split of data and the split of histogram is consistent. See the `CheckSplitValid` method in cuda_single_gpu_tree_learner.cpp.

File: src/treelearner/cuda/cuda_data_partition.cu (hunk 3 above)
Comment (shiyu1994) [reply src/treelearner/cuda/cuda_data_partition.cu:1048]: No. When USE_BAGGING is false, `data_indices_in_leaf` will be all the data indices of training data. So `local_data_index` is exactly `global_data_index`.

File: src/metric/metric.cpp (hunk 4 above)
Comment (shiyu1994) [reply src/metric/metric.cpp:140]: In the next stage, we need to merge the CUDA version of the metrics one by one. I think writing explicitly for each metric and for both `cuda_exp` and other versions is more convenient for the next stage.

**Comment by shiyu1994** [comment]
> > That will make the single-GPU version of `cuda_exp` complete.
> 
> I think you forgot about
> 
> * [[CUDA] Enable use of subset Dataset in new CUDA version #5086](https://github.com/microsoft/LightGBM/issues/5086)
> * [[CUDA] Monotonic Constraints on New CUDA Version #4781](https://github.com/microsoft/LightGBM/issues/4781)
> * [[CUDA] Add support for min_data_in_group in new CUDA version #5062](https://github.com/microsoft/LightGBM/issues/5062)
> * [[CUDA] Feature Interaction Constraints on New CUDA Version #4785](https://github.com/microsoft/LightGBM/issues/4785)

Thanks for the reminder. Definitely we also need those features to make our new CUDA version complete.

**Comment by shiyu1994** [comment]
Close and reopen to trigger the ci.

**Comment by shiyu1994** [comment]
Close and reopen to trigger ci tests.

**Review by guolinke — APPROVED** [review]
LGTM
